### PR TITLE
Clean and unify the Jenkinsfile

### PR DESCRIPTION
**Pull request by emanuelez**
It also adds Slack notifications for the `master` and `releases` branches and introduces lint checking.
It doesn't fail be build yet since we have 135 issues to solve (in our unit tests)

### CODE REVIEW

**Comment by emanuelez** [comment]
@realm/java @teotwaki @alebsack

File: Jenkinsfile [hunk 1]
```diff
@@ -1,169 +1,158 @@
...
-        def rootFolder = new File('unzipped/jni')
-        rootFolder.traverse (type: DIRECTORIES) { folder ->
-            def abiName = folder.name()
-            def libSize = new File(folder, 'librealm-jni.so').size() as String
-            sendTaggedMetric('abi_size', libSize, 'type', abiName)
-        }
+  dir('realm/realm-library/build/outputs/aar') {
+    sh '''set -xe
+    unzip realm-android-library-release.aar -d unzipped
+    find $ANDROID_HOME -name dx | sort -r | head -n 1 > dx
```
Comment (kneth) [inline Jenkinsfile]: Why store in a file? An environment variable might be an alternative.

**Comment by kneth** [comment]
It looks fine to me - but it's a bit out of my typical range ;-)

File: Jenkinsfile (hunk 1 above)
Comment (emanuelez) [inline Jenkinsfile]: because that's the only way to pass data from a `sh` section back into Groovy. Environment variables are ephemeral in this context.

Comment (zaki50) [inline Jenkinsfile]: but it doesn't seem to be used in Groovy.

Comment (emanuelez) [inline Jenkinsfile]: Ah right! That's is old code from Christian. I didn't touch that part (apart from the indentation). Agreed, a variable could do, but it's a bit out of the scope of the PR

File: Jenkinsfile [hunk 2]
```diff
@@ -1,169 +1,158 @@
...
+
+def buildSuccess = false
+try {
+  node('docker') {
+    // Allocate a custom workspace to avoid having % in the path (it breaks ld)
+    ws('/tmp/realm-java') {
+      stage 'SCM'
+      checkout scm
+      // Make sure not to delete the folder that Jenkins allocates to store scripts
+      sh 'git clean -ffdx -e .????????'
```
Comment (cmelchior) [inline Jenkinsfile]: Could we narrow the regular expression a bit? ie. to only alphanumerical characters or something?

File: Jenkinsfile [hunk 3]
```diff
@@ -1,169 +1,158 @@
...
-            'archive': true,
-            'glob' : 'logcat.txt'
+        // TODO: add support for running monkey on the example apps
+
+        if (env.BRANCH_NAME == 'master') {
+          stage 'Collect metrics'
+          collectAarMetrics()
+
+          stage 'Publish to OJO'
+          gradle 'assemble ojoUpload'
```
Comment (cmelchior) [inline Jenkinsfile]: should this be `gradle('assemble ojoUpload') instead ?

File: Jenkinsfile (hunk 2 above)
Comment (emanuelez) [inline Jenkinsfile]: I wish we had regexes! :( Only shell wildcards are allowed here

File: Jenkinsfile (hunk 3 above)
Comment (emanuelez) [inline Jenkinsfile]: In Groovy if you only have one argument you can omit the brackets and it's used all over the place in the Jenkinsfile DSL as you can see

**Comment by alebsack** [comment]
:+1 Although my groovy isn't great yet...  I'll be sure to track this repo for examples. :)

File: Jenkinsfile [hunk 4]
```diff
@@ -1,169 +1,158 @@
...
-            }
+import groovy.json.JsonOutput
+
+def buildSuccess = false
+try {
+  node('docker') {
+    // Allocate a custom workspace to avoid having % in the path (it breaks ld)
+    ws('/tmp/realm-java') {
+      stage 'SCM'
+      checkout scm
```
Comment (radu-tutueanu) [inline Jenkinsfile]: Does `checkout scm` use the parameters defined in the Jenkins job SCM stage?

**Comment by radu-tutueanu** [comment]
👍 Looks good. I will be using this as a reference for groovy scripts as well.

File: Jenkinsfile (hunk 4 above)
Comment (emanuelez) [inline Jenkinsfile]: Yes, and in this case it's completely handled by the Github Organisation Folder plugin

**Pull request merged**